File: apps/event-consumer/src/ticket-analysis/algorithm/find-closest-km.js
import SplitKM from '../split-km.js';
import findClosestNumber from './find-closest-number.js';
import { kmRetrieve } from './km.js';

/**
 * @param {SplitKM} km
 * @param {SplitKM[]} listOfKm
 * @returns {SplitKM?}
 */
const tryFindExactSecondNumber = (km, listOfKm) => {
  const secondNumber = km.parts[1] || 0;
  for (const km of listOfKm) {
    const itemSecondNumber = km.parts[1] || 0;
    if (
      itemSecondNumber === secondNumber ||
      itemSecondNumber === Number(`${secondNumber}0`) ||
      itemSecondNumber === Number(`${secondNumber}00`)
    )
      return km;
  }
  return null;
};

/**
 * @param {string} kmTarget
 * @param {string[]} kmList
 */
const findIndex = (kmTarget, kmList) => {
  if (kmList.includes(kmTarget)) {
    return kmList.indexOf(kmTarget);
  }

  const kmTargetSplit = SplitKM.valueOf(kmTarget);
  const listOfKMSplit = kmList.map((i) =>
    i.toLowerCase().includes('km')
      ? SplitKM.valueOf(kmRetrieve(i))
      : SplitKM.valueOf(i)
  );

  const firstPartClosestNumber = findClosestNumber(
    listOfKMSplit.map((i) => i.parts[0]),
    kmTargetSplit.parts[0]
  );
  const listFilteredWithFirstPartClosestNumber = listOfKMSplit.filter(
    (i) => i.parts[0] === firstPartClosestNumber
  );
  if (listFilteredWithFirstPartClosestNumber.length === 1) {
    return kmList.indexOf(listFilteredWithFirstPartClosestNumber[0].source);
  }

  const secondPartOfKmNumber = kmTargetSplit.parts[1] || null;
  if (secondPartOfKmNumber == null) {
    const result = listFilteredWithFirstPartClosestNumber.find(
      (i) => i.parts.length === 1
    );
    if (!result) {
      const firstItemFromListFiltered =
        listFilteredWithFirstPartClosestNumber[0];
      return kmList.indexOf(firstItemFromListFiltered.source);
    }
    return kmList.indexOf(result.source);
  }

  const resultFromSecondNumber = tryFindExactSecondNumber(
    kmTargetSplit,
    listFilteredWithFirstPartClosestNumber
  );
  if (resultFromSecondNumber)
    return kmList.indexOf(resultFromSecondNumber.source);

  const secondPartOfKmNumberClosest = findClosestNumber(
    listFilteredWithFirstPartClosestNumber.map((i) => i.parts[1] || 0),
    secondPartOfKmNumber
  );
  const result = listFilteredWithFirstPartClosestNumber.find((i) => {
    return (
      (i.parts[0] === firstPartClosestNumber && i.parts[1]) ||
      secondPartOfKmNumberClosest === 0
    );
  });
  if (!result) {
    throw new Error('Result not found');
  }
  return kmList.indexOf(result.source);
};

export default findIndex;

File: core/src/main/kotlin/br/com/cstag/core/services/analyze/utils/FindClosestKm.kt
package br.com.cstag.core.services.analyze.utils

object FindClosestKm {
    fun findIndex(kmTarget: String, listOfKm: List<String>): Int {
        val kmTargetSplit = SplitKm.valueOf(kmTarget)
        val listOfKmSplit = listOfKm.map {
            try {
                SplitKm.valueOf(it)
            } catch (e: Exception) {
                SplitKm.valueOf(KmNumberRetriever.retrieve(it))
            }
        }.sort()

        val firstPartClosestNumber =
            FindClosestNumber.findClosest(listOfKmSplit.map { it.parts[0] }, kmTargetSplit.parts[0])
        val listFilteredWithFirstPartClosestNumber = listOfKmSplit.filter { it.parts[0] == firstPartClosestNumber }

        val secondPartOfKmNumber = kmTargetSplit.parts.getOrNull(1)
        if (secondPartOfKmNumber == null) {
            val result = listFilteredWithFirstPartClosestNumber.find { it.parts.size == 1 }
            if (result == null) {
                val firstItemFromListFiltered = listFilteredWithFirstPartClosestNumber[0]
                return listOfKm.indexOf(firstItemFromListFiltered.source)
            }
            return listOfKm.indexOf(result.source)
        }

        tryFindExactSecondNumber(kmTargetSplit, listFilteredWithFirstPartClosestNumber)?.let {
            return listOfKm.indexOf(it.source)
        }

        val secondPartOfKmNumberClosest = FindClosestNumber.findClosest(
            listFilteredWithFirstPartClosestNumber.map { it.parts.getOrElse(1) { 0 } }, secondPartOfKmNumber
        )
        val result = listFilteredWithFirstPartClosestNumber.find {
            it.parts[0] == firstPartClosestNumber && it.parts.getOrElse(1) {0} == secondPartOfKmNumberClosest
        } ?: throw Error("Result not found")
        return listOfKm.indexOf(result.source)
    }

    private fun tryFindExactSecondNumber(km: SplitKm, listOfKm: List<SplitKm>): SplitKm? {
        val secondNumber = km.parts.getOrElse(1) {0}
        listOfKm.forEach {
            val itemSecondNumber = it.parts.getOrElse(1) {0}
            if ((itemSecondNumber == secondNumber)
                || (itemSecondNumber == "${secondNumber}0".toInt())
                || (itemSecondNumber == "${secondNumber}00".toInt()))
                    return it
        }
        return null
    }
}